### ISSUE TITLE
Recent change to `Matrix` seriously breaks the `Semigroups` package

### ISSUE THREAD

**Issue by james-d-mitchell** [gap-system/gap#4814]
### Observed behaviour

~~~

#I  Running tests with acting methods enabled
Semigroups package: testinstall.tst
msecs: [45](https://github.com/semigroups/Semigroups/runs/5424948669?check_suite_focus=true#step:6:45)49
#I  Running tests with acting methods disabled
Semigroups package: testinstall.tst
msecs: 3256
#I  Running tests with acting methods enabled
Architecture: x86_64-apple-darwin20.6.0-default64-kv8

testing: /Users/runner/work/Semigroups/Semigroups/gaproot/pkg/Semigroups/tst/s\
tandard/attributes/acting.tst
     341 ms (209 ms GC) and [48](https://github.com/semigroups/Semigroups/runs/5424948669?check_suite_focus=true#step:6:48).2MB allocated for attributes/acting.tst
testing: /Users/runner/work/Semigroups/Semigroups/gaproot/pkg/Semigroups/tst/s\
tandard/attributes/attr.tst
######## > Diff in:
/Users/runner/work/Semigroups/Semigroups/gaproot/pkg/Semigroups/tst/standard/a\
ttributes/attr.tst:1215
# Input is:
S := Monoid(Matrix(GF(2 ^ 2),
[[Z(2 ^ 2), 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2)],
 [Z(2 ^ 2), 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2)],
 [0 * Z(2), Z(2) ^ 0, 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2)],
 [0 * Z(2), 0 * Z(2), Z(2) ^ 0, 0 * Z(2), 0 * Z(2), 0 * Z(2)],
 [0 * Z(2), 0 * Z(2), 0 * Z(2), Z(2) ^ 0, 0 * Z(2), 0 * Z(2)],
 [0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2), 0 * Z(2)]]));;
# Expected output:
# But found:
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 2nd choice method found for `OneMutable' on 1 arguments
~~~

### Expected behaviour

It should work, as it did before 3e9fb2ab45457f97807bea4762358dc0489adf9a was merged. According to the comment in the code in the commit:

~~~
# rank higher than a method in the semigroups package, which otherwise jumps
# in and causes an error when testing
# line 318 of semigroups-3.4.0/gap/elements/semiringmat.gi
~~~

this was known at the time the code was written. 

### Copy and paste GAP banner (to tell us about your setup)

~~~
┌───────┐   GAP 4.12dev built on 2022-03-04 16:16:47+0000
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-apple-darwin20.6.0-default64-kv8
 Configuration:  gmp 6.2.1, GASMAN
 Loading the library and packages ...
 Packages:   AClib 1.3.2, Alnuth 3.1.2, AtlasRep 2.1.0, AutoDoc 2020.08.11, 
             AutPGrp 1.10.2, CRISP 1.4.5, Cryst 4.1.24, CrystCat 1.1.9, 
             CTblLib 1.3.3, FactInt 1.6.3, FGA 1.4.0, Forms 1.2.6, 
             GAPDoc 1.6.5, genss 1.6.6, IO 4.7.2, IRREDSOL 1.4.3, 
             LAGUNA 3.9.3, orb 4.8.4, Polenta 1.3.9, Polycyclic 2.16, 
             PrimGrp 3.4.1, RadiRoot 2.8, recog 1.3.2, ResClasses 4.7.2, 
             SmallGrp 1.4.2, Sophus 1.24, SpinSym 1.5.2, TomLib 1.2.9, 
             TransGrp 3.3, utils 0.72
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
~~~

**Comment by james-d-mitchell** [comment]
@hulpke

**Comment by james-d-mitchell** [comment]
I can fix the particular issue that is shown above, but basically https://github.com/gap-system/gap/commit/3e9fb2ab45457f97807bea4762358dc0489adf9a completely breaks the stuff in the `Semigroups` package for matrices over finite fields.

**Comment by fingolfin** [comment]
@james-d-mitchell Thanks for reporting. It's unfortunate that the package CI reporting is broken (and has been for years) and we were not yet able to re-establish it during the recent GAP Days, otherwise we would have noticed this sooner.

Anyway, we'll work it out one way or another, definitely before GAP gets a release.

**Comment by hulpke** [comment]
The problem is that the semigroup package installs a method for the library operation `Matrix` for arguments: [ (semi)ring and homogeneous list] that enforces (with an error message) that matrices defined over rings must be defined over finite fields or must be defined over the integers.
[The error was thrown up by standard tests if the semigroups package is loaded]

I am not sure why semigroup does this. Does it use the `Matrix` command for something different than IsMatrixObj? Or why does it need to catch the creation of matrices over other rings?

**Comment by james-d-mitchell** [comment]
> The problem is that the semigroup package installs a method for the library operation `Matrix` for arguments:

The function `Matrix` has existed in the released version of the Semigroups package since 2017 (and in the dev version since 2015), which is before matrix objects were being used widely in GAP, and the method you are referring to (in the Semigroup package) was the only one for creating matrices over semirings until recently. We didn't use `MatrixObj`s at the time because they didn't really exist, and their future wasn't clear. 

> that enforces (with an error message) that matrices defined over rings must be defined over finite fields or must be defined over the integers.

This is not the case, if a `Matrix` is not defined over the integers or a finite field, then the method you refer to calls `TryNextMethod`.  I suppose the error message arises because of the fact that a `MatrixObj` is not returned. 

The basic point here is that your recent changes seriously break the `Semigroups` package code that has worked since at least 2017, fixing this by making changes to the `Semigroups` package would require a lot of work, or will break backwards compatibility, and this doesn't seem to have even been considered (never mind discussed or agreed) prior to merging your code. Given that the GAP tests failed when the Semigroups package method was used, it wouldn't have been much of a stretch to realise that the opposite would also be true, but the merge took place anyway...

As I've said elsewhere I'm happy to come to an agreement/compromise that will resolve this (possibly including one where we make significant changes to the Semigroups package), and hope that we can have a more informed discussion about the issue.

**Comment by hulpke** [comment]
OK, I've finally managed to install semigroup. What happens is that (in the character table code) one has a list of plist vectors:
```
brk> D.modulars;
[ <plist vector over GF(421) of length 5> ]
```
and wants to build a matrix from it:
```
Matrix(GF(421),D.modulars);
```
This causes (if `semigroup` is loaded) an error message:
```
Error, Semigroups: Matrix: usage,
the 2nd argument <mat> does not give a rectangular table,
```
The method ranking ensures that instead of the "semigroup" method for `Matrix` which triggers the error, instead the library method is called.

If you want to reproduce this, change the ranking (line 303 in matobj.gi) back to 0. Load `semigroup`. Then call
```
gap> Irr(AlternatingGroup(5));
```

**Comment by james-d-mitchell** [comment]
Thanks @hulpke. Equally, if you load the Semigroups package with the master branch of GAP, and run:

~~~
SemigroupsTestStandard();
~~~

You'll be able to reproduce the bug that I posted in the issue above.

**Comment by hulpke** [comment]
Ultimately, the underlying question is on the definition of `Matrix` between library and packages. I'm probably not the right person to make that call.

As a workaround, if the `Matrix` method in the semigroups package called `TryNextMethod();` instead of forcing an error, we could undo the forced ranking of the library method (which hopefully would remove the problem at hand).

**Comment by fingolfin** [comment]
@james-d-mitchell I guess this is resolved by Semigroups 5.0.0? At least in the sense that things work for now (we still want to improve things, of course)

**Comment by james-d-mitchell** [comment]
Yup, fully resolved!
